Allow setting credentials info via env vars

justinstoller opened the pull request:
Prior to this a user could set the credential information (username,
password, and url of the forge) via a file located at
~/.puppetforge.yml.

This commit allows users to set the environment variables
 - BLACKSMITH_FORGE_URL
 - BLACKSMITH_FORGE_USERNAME
 - BLACKSMITH_FORGE_PASSWORD
to set the url, username, and password values respectively.

This eases publishing under multiple usernames on the forge as well as
pushing to multiple forges by users (and CI systems) by allowing a
method of setting variables that is not per user wide (an issue with a
credentials file in the user's home directory).

Environment variables will take precedence over values set in a
credentials file.

DavidS commented:
:+1:

nibalizer commented:
This looks good to me. I'm not sure why it has (maint) on it because this seems like a feature.
I'd like to see some docs in the README as well.

justinstoller commented:
I'll go ahead and add some docs.

The (maint) is because the work doesn't have an internal ticket, but is something I did in some free time (is that like corporate PTSD?). Looking at the previous few commits I don't see a pattern to how their prefaced, so I'll just remove the (maint) unless you have a recommendation for how it should be worded.

flosell commented:
:+1:

flosell commented on lib/puppet_blacksmith/forge.rb:
```diff
@@ -65,27 +62,71 @@ def push!(name, package = nil)
+      file = File.expand_path(CREDENTIALS_FILE)
+
+      if File.exists?(file)
+        credentials = YAML.load_file(CREDENTIALS_FILE)
```
shouldn't this be `YAML.load_file(file)`? 

I just tried your path with environment variables set and the credentials file still there and got this: 
```
Errno::ENOENT: No such file or directory @ rb_sysopen - ~/.puppetforge.yml
/var/opt/jenkins/.gem/ruby/2.1.5/gems/puppet-3.8.5/lib/puppet/vendor/safe_yaml/lib/safe_yaml.rb:188:in `initialize'
/var/opt/jenkins/.gem/ruby/2.1.5/gems/puppet-3.8.5/lib/puppet/vendor/safe_yaml/lib/safe_yaml.rb:188:in `open'
/var/opt/jenkins/.gem/ruby/2.1.5/gems/puppet-3.8.5/lib/puppet/vendor/safe_yaml/lib/safe_yaml.rb:188:in `unsafe_load_file'
/var/opt/jenkins/.gem/ruby/2.1.5/gems/puppet-3.8.5/lib/puppet/vendor/safe_yaml/lib/safe_yaml.rb:153:in `load_file_with_options'
/var/opt/jenkins/.gem/ruby/2.1.5/bundler/gems/puppet-blacksmith-92151cf9cc73/lib/puppet_blacksmith/forge.rb:106:in `load_credentials_from_file'
/var/opt/jenkins/.gem/ruby/2.1.5/bundler/gems/puppet-blacksmith-92151cf9cc73/lib/puppet_blacksmith/forge.rb:66:in `load_credentials'
/var/opt/jenkins/.gem/ruby/2.1.5/bundler/gems/puppet-blacksmith-92151cf9cc73/lib/puppet_blacksmith/forge.rb:19:in `initialize'
/var/opt/jenkins/.gem/ruby/2.1.5/bundler/gems/puppet-blacksmith-92151cf9cc73/lib/puppet_blacksmith/rake_tasks.rb:75:in `new'
/var/opt/jenkins/.gem/ruby/2.1.5/bundler/gems/puppet-blacksmith-92151cf9cc73/lib/puppet_blacksmith/rake_tasks.rb:75:in `block (2 levels) in define'
```
When I remove the credentials file it works

justinstoller commented:
I added a mention in the README, updated the tests to catch the issue that @flosell found, and then fixed the issue itself. I squashed it back into one commit and then removed the "(maint)" in the commit message (and PR title). Though now I have to figure out why I didn't catch the issue earlier in my environment....

Apologies for the delay.

jyaworski commented:
This LGTM.
